File: CALC_NUM/parte2-equacoes_algebricas_e_transcendentes/bissecao.cpp
#include <iostream>
#include <iomanip>
#include <cmath>

using namespace std;

long double calcular_xm(long double a, long double b) {
    return (a + b) / 2;
}

long double funcao(long double x) {
    return exp(x) - (2 * cos(x));
}

int main(void) {
    const double erro_admitido = pow(10, -15);

    long double a = 0;
    long double b = 2;
    long double xm = calcular_xm(a, b);
    long double erro = abs(a - b);

    long double funcao_a, funcao_xm;
    while (erro > erro_admitido) {
        xm = calcular_xm(a, b);
        funcao_a = funcao(a);
        funcao_xm = funcao(xm);

        if (funcao_xm == 0) {
            return xm;
        }

        if (funcao_a * funcao_xm < 0) {
            b = xm;
        } else {
            a = xm;
        }

        erro = abs(a - b);
    }

    cout << setprecision(15);
    cout << "Valor de xm: " << xm << endl;
    cout << " Erro: " << erro << endl;

    return 0;
}

File: CALC_NUM/parte2-equacoes_algebricas_e_transcendentes/bissecao.py
from math import exp, cos

erro_admitido = 10 ** -15
# valores iniciais das variaveis
a = 0
b = 2
xm = (a + b) / 2
erro = abs(a - b)


def calcular_xm(a: float, b:float) -> float:
    return (a + b) / 2


def funcao(x: float) -> float: 
    return exp(x) - (2 * cos(x))


while erro > erro_admitido:
    xm = calcular_xm(a, b) 
    funcao_a = funcao(a)
    funcao_xm = funcao(xm)
    
    if (funcao_a * funcao_xm) < 0: 
        b = xm
    else:
        a = xm
    
    erro = abs(a - b)

print(xm)
